README.md
![](docs/img/majortom_banner.png)

Navigate your file system from the command line using shortcuts.

### List shortcuts
Use `to` (with no arguments) to list your current shortcuts.
![](docs/img/majortom_get_shortcuts.png)

### Navigate
Use `to <shortcut>` to navigate (change directory) to an existing shortcut.

You can also abbreviate the shortcut name (e.g. `to apa` for `apache`).
![](docs/img/majortom_cd_to_apache.png)

### See where you are
When you run `to` (with no arguments) all shortcuts matching your current directory will be marked.

![](docs/img/majortom_show_current.png)
### Add a shortcut
Use `to -a <shortcut>` to add a shortcut to your current working directory.
![](docs/img/majortom_add_shortcut.png)

### Delete a shortcut
Use `to -d <shortcut>` to delete an existing shortcut.
![](docs/img/majortom_delete_shortcut.png)


### Supported Color Modes
MajorTom defaults to terminal color mode `16m` (24-bit RGB). It also supports
`256`,`16`, and `no-color` modes via the following command-line options.  

- `-color=16m`
- `-color=256`
- `-color=16`
- `-no-color`

If you want to force a color mode other than the default (`16m`), then add it to the `to()` function
in your `.bashrc`/`.zshrc` like this:

```bash
to () {
    result=$(majortom -color=16 $@)
...
```

![](docs/img/majortom_colors.png)

### Supported Shells
MajorTom supports `zsh` and `bash`.  During installation it will look for `~/.zshrc` and `~/.bashrc` 
and will prompt you to install its helper function `to()` into each (if found).

To `to()` function is extremely simple, so if you're running a different shell and have some basic
familiarity with shell scripting then you should have no trouble porting it to a different shell.

## Installation

### Installing From a Pre-Made Build
- Download one of the pre-made builds from [Releases](https://github.com/epmoyer/majortom/releases).
- Extract the archive
    - For tar archives: `tar -xvzf <archive>` e.g. `tar -xvzf majortom_1.4.0.linux.arm.tgz`
- `cd` into the extracted archive.
- run `./install.sh`

#### Installation example
![](docs/img/majortom_install.png)

### Installing From Source
- Follow the [Build Instructions](#build-instructions) to build the installation images.
- `cd` into the build for your OS and processor architecture under `dist/builds`
- Run `./install.sh`

## Uninstalling
- Delete the `to()` helper function from your `~/.bashrc` and/or `~/.zshrc`.
    - Delete the lines from:
        -  `# majortom:start -----------------------------------------------------------------` 
    - Through:
        - `# majortom:end -----------------------------------------------------------------`
- Delete the executable:
    - `sudo rm /usr/local/bin/majortom`

## Configuration File
By default, MajorTom's config file location defaults to `~/.config/majortom/majortom_config.json`.  You can override that by setting the environment variable `MAJORTOM_CONFIG`.  (e.g. `export MAJORTOM_CONFIG=~/my_config_dir/mt.json`).

If you don't yet have a config file, you can create one by running `majortom -init`, which will create a new (blank) config file at the currently configured location.

The `-init` command will never erase/overwrite/clear an existing config file.

A typical config file containing a few shortcuts might look like this:

```json
{
    "locations": {
        "apache": "/var/log/apache2",
        "dne": "~/this/path/does/not/exist",
        "hello": "~/code/golang/helloworld",
        "launch": "~/Library/Preferences/com.apple.LaunchServices",
        "my": "~/code/golang/myproject"
    }
}
```

## Build Instructions
_(These instructions presume that you have a working [go](https://go.dev) environment set up on your machine)_

### Building from source
- Clone the repo.
- `cd` into the repo.
- Run `go get .` to fetch the dependencies.
- Run `./build.sh`
    - The builds will be created in `/dist/builds`
    - the `zip`/`tar` images will be created in `/dist/images`

### Building for other architectures/processors
The build script creates cross-compiled builds for multiple OS's and processor architectures.  If you need to create a build for different OS or architecture, add a new one to the end of the `.build.sh` script.

For example, this line creates a release for the `linux` OS and the `amd64` architecture, and packages the output as a `tar` image:

`do_build linux amd64 tar`

The API of the `do_build()` function is basically `do_build(GOOS, GOARCH, [tar|zip])` so you can pass any `GOOS` or `GOARCH` that `go` [supports](https://go.dev/doc/install/source#environment).

**NOTE**: For clarity, if the `GOOS` is set to `darwin` then the name `macos` gets used when creating build outputs. 

## How it works
The `to` command (defined in `/dist/resources/shell_init_snippet.sh`) is a shell script function which gets added to your shell init script (`.bahsrc`, `.zshrc`).

`to()` calls `majortom` (which gets installed in `/usr/local/bin/majortom`), and if `majortom` returns a path (which it will prefix with `:`) then `to` will `cd` to it.

You can see this behavior if you run `majortom` directly...
![](docs/img/majortom_direct_execution_apache.png)

...versus running it via `to`...
![](docs/img/majortom_cd_to_apache.png)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.4.0 - 2023-01-16
### Changed
- `-init` will create parent directories if they don't exist.
    - NOTE: for a brief time there was a `-force` option (to force `-init` without prompting) but it turned out that prompts were not visible to thn user because thy occur within the (output capturing) shell function to(). So now `-init` creates the file (and paths) unconditionally, so long as the config file does not already exist.t

## 1.3.0 - 2023-01-14
### Added
- Show `(DOES NOT EXIST)` for paths that do not exist if in `-no-color` mode.

## 1.2.0 - 2023-01-14
### Added
- Color options `-color=` (`16`, `256`, `16m`) and `-no-color`
### Changed
- Use `gtar` for tar archive creation, if available

## 1.1.0 - 2023-01-06
### Changed
- Adopt app name `majortom`

## 1.0.0 - 2023-01-06
Initial Releare

to.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"os/user"
	"path/filepath"
	"sort"
	"strings"

	"github.com/gookit/color"
)

const APP_NAME = "to"
const APP_VERSION = "1.0.0"

type ConfigDataT struct {
	Locations map[string]string `json:"locations"`
}

var styleShortcut = color.HEXStyle("#ff8000")
var stylePath = color.HEXStyle("#00ffff")
var stylePathDNE = color.HEXStyle("#808080")
var styleCurrent = color.HEXStyle("#ffff00")
var styleError = color.HEXStyle("#ff4040")

var EXIT_CODE_SUCCESS = 0
var EXIT_CODE_FAIL = 1

var ENV_VAR_CONFIG = "TO_CONFIG_DB"

var CONFIG_FILENAME = "to_shortcuts.json"

func main() {
	flag.Usage = func() {
		w := flag.CommandLine.Output() // may be os.Stderr - but not necessarily
		fmt.Fprintf(w, "Usage of %s:\n", os.Args[0])
		flag.PrintDefaults()
		fmt.Fprintf(w,
			"NOTE:\n   Set the environment variable %s to point to to's configuration json file.\n",
			ENV_VAR_CONFIG)
		fmt.Fprintf(w, "   Currently %s = \"%s\"\n", ENV_VAR_CONFIG, os.Getenv(ENV_VAR_CONFIG))
	}

	optVersion := flag.Bool("version", false,
		"Show version.")
	optAdd := flag.Bool("a", false,
		"Add current path (as requested shortcut).")
	optDelete := flag.Bool("d", false,
		"Delete requested shortcut.")
	flag.Parse()

	if *optVersion {
		fmt.Printf("%s %s\n", APP_NAME, APP_VERSION)
		os.Exit(EXIT_CODE_SUCCESS)
	}

	args := flag.Args()
	if len(args) > 1 {
		// Too many args.
		fmt.Println("Too many arguments.")
		os.Exit(EXIT_CODE_FAIL)
	}

	config := loadConfig()

	if len(args) == 0 {
		showShortcuts(config)
		os.Exit(EXIT_CODE_SUCCESS)
	}
	if *optAdd {
		config = addShortcut(config, args[0])
		saveConfig(config)
		os.Exit(EXIT_CODE_SUCCESS)
	}
	if *optDelete {
		config = deleteShortcut(config, args[0])
		saveConfig(config)
		os.Exit(EXIT_CODE_SUCCESS)
	}
	path := getPath(config, args[0])
	path = expandHome(path)
	fmt.Printf(":%s\n", path)

	os.Exit(EXIT_CODE_SUCCESS)
}

func addShortcut(config ConfigDataT, shortcut string) ConfigDataT {
	currentPath, err := os.Getwd()
	if err != nil {
		fmt.Println(err)
		os.Exit(EXIT_CODE_FAIL)
	}
	currentPath = abbreviateHome(currentPath)
	config.Locations[shortcut] = currentPath
	fmt.Printf("Adding shortcut \"%s\"...\n", shortcut)
	return config
}

func deleteShortcut(config ConfigDataT, shortcut string) ConfigDataT {
	if _, ok := config.Locations[shortcut]; ok {
		delete(config.Locations, shortcut)
	} else {
		styleError.Printf(
			"Shortcut \"%s\" does not exist.\n",
			shortcut)
		os.Exit(EXIT_CODE_FAIL)
	}
	fmt.Printf("Deleting shortcut \"%s\"...\n", shortcut)
	return config
}

func getPath(config ConfigDataT, shortcut string) string {
	paths := make([]string, 0)
	matched_keys := make([]string, 0)

	for key, path := range config.Locations {
		if key == shortcut {
			// Always return the path if the key EXACTLY matches the requested shortcut
			return path
		}
		if strings.HasPrefix(key, shortcut) {
			paths = append(paths, path)
			matched_keys = append(matched_keys, key)
		}
	}
	if len(paths) == 0 {
		styleError.Printf(
			"No match found for shortcut \"%s\". Run \"to\" with no arguments for a list of shortcuts.\n",
			shortcut)
		os.Exit(EXIT_CODE_FAIL)
	}
	if len(paths) > 1 {
		message := "Matched multiple shortcuts: "
		for i, key := range matched_keys {
			message += styleShortcut.Sprintf("%s", key)
			if i < len(matched_keys)-1 {
				message += ", "
			}
		}
		fmt.Println(message)
		os.Exit(EXIT_CODE_FAIL)
	}
	// Return the (single) matching path
	return paths[0]
}

func showShortcuts(config ConfigDataT) {
	currentPath, err := os.Getwd()
	if err != nil {
		fmt.Println(err)
		os.Exit(EXIT_CODE_FAIL)
	}

	var maxLen int
	shortcuts := make([]string, 0)
	for shortcut, _ := range config.Locations {
		if len(shortcut) > maxLen {
			maxLen = len(shortcut)
		}
		shortcuts = append(shortcuts, shortcut)
	}
	fmt.Println("Available shortcuts:")
	sort.Strings(shortcuts)
	for _, shortcut := range shortcuts {
		path := config.Locations[shortcut]
		pathAbsolute := expandHome(path)
		if pathAbsolute == currentPath {
			styleCurrent.Printf("▶ %-*s ", maxLen, shortcut)
		} else {
			styleShortcut.Printf("  %-*s ", maxLen, shortcut)
		}
		if _, err := os.Stat(pathAbsolute); !os.IsNotExist(err) {
			// Path exists
			stylePath.Printf("%s", path)
		} else {
			// Path does not exist
			stylePathDNE.Printf("%s", path)
		}
		// We cannot include the \n in the final style .Printf() above, otherwise the \n occurs
		// before the subsequent escape code for clearing the style, which causes an additional
		// linefeed to be printed when we echo the output text in bash/zsh.
		fmt.Print("\n")
	}
}

func expandHome(path string) string {
	usr, _ := user.Current()
	homeDir := usr.HomeDir

	if path == "~" {
		// In case of "~", which won't be caught by the "else if"
		return homeDir
	} else if strings.HasPrefix(path, "~/") {
		// Use strings.HasPrefix so we don't match paths like
		// "/something/~/something/"
		return filepath.Join(homeDir, path[2:])
	}
	return path
}

func abbreviateHome(path string) string {
	usr, _ := user.Current()
	dir := usr.HomeDir
	return strings.Replace(path, dir, "~", 1)
}

func loadConfig() ConfigDataT {
	pathConfig := getConfigPath()
	jsonFile, err := os.Open(pathConfig)
	if err != nil {
		fmt.Println(err)
		os.Exit(EXIT_CODE_FAIL)
	}
	defer jsonFile.Close()

	var shortcuts ConfigDataT
	byteValue, _ := ioutil.ReadAll(jsonFile)
	json.Unmarshal(byteValue, &shortcuts)
	return shortcuts
}

func saveConfig(config ConfigDataT) {
	pathConfig := getConfigPath()
	file, _ := json.MarshalIndent(config, "", "    ")
	_ = ioutil.WriteFile(pathConfig, file, 0644)
}

func getConfigPath() string {
	path := os.Getenv(ENV_VAR_CONFIG)
	if path == "" {
		styleError.Printf(
			"Environment variable %s is not set. Set it to the path of to's config json.\n", ENV_VAR_CONFIG)
		os.Exit(EXIT_CODE_FAIL)
	}
	return path
}


build.sh
#/bin/bash
GREEN=$'\033[32m'
CYAN=$'\033[36m'
ENDCOLOR=$'\033[0m'

# Extract the app name and version number from the main go file by brute force
APP_VERSION=`awk '/^const APP_VERSION/{print $4}' majortom.go | sed -e 's/"//g'`
APP_NAME=`awk '/^const APP_NAME/{print $4}' majortom.go | sed -e 's/"//g'`

echo "Building $APP_NAME version $APP_VERSION..."

if ! command -v gtar &> /dev/null
then
    TAR_APP=tar
    echo "gtar not found. Will use tar instead."
    echo "NOTE: If you are building on MacOS then the native MacOS tar will cause"
    echo "      (benign) warnings about the extended header keyword LIBARCHIVE.xattr.com.dropbox.attrs"
    echo "      when un-tar-ing the resulting tar archive on Linux systems.  Using gtar avoids this."
else
    TAR_APP=gtar
    echo "gtar found. Will use gtar."
fi
# TAR_APP=tar

do_build () {
    USE_GOOS=$1
    USE_GOARCH=$2
    IMAGE_TYPE=$3

    OS_NAME=$USE_GOOS
    if [[ $OS_NAME = "darwin" ]]
    then
        # We will use the name 'macos' instead of 'darwin' for build images to provide clarity
        # for a broader range of macos users.
        OS_NAME='macos'
    fi

    TARGET_NAME=$OS_NAME.$USE_GOARCH

    BUILD_DIR_BASE=dist/builds
    BUILD_DIR_FINAL=$APP_NAME\_$APP_VERSION.$TARGET_NAME
    TARGET_DIR=$BUILD_DIR_BASE/$BUILD_DIR_FINAL

    echo "${CYAN}$USE_GOOS:$USE_GOARCH${ENDCOLOR} -------------------------------------------"

    echo "Building into: $TARGET_DIR..."

    echo "Making directory (if missing): $TARGET_DIR..."
    mkdir -p $TARGET_DIR

    # Build executable
    GOOS=$USE_GOOS GOARCH=$USE_GOARCH go build -o $TARGET_DIR/majortom

    # Copy supporting files
    cp dist/resources/install.sh $TARGET_DIR
    cp dist/resources/shell_init_snippet.sh $TARGET_DIR

    # I don't really like the non-determinism of this sleep command, but without it the tar below
    # would sometimes throw the warning "gtar: ./majortom: file changed as we read it".
    # Presumably the go build is backgrounding some stage of operation, or the file system is
    # synching.  I tried 'sync' and 'wait', but only `sleep 1` appears to (for now) have reliably
    # squashed the warning (which is really an error to us, if the build has not completed before
    # we tar).
    sleep 1

    # Build release
    if [ "$IMAGE_TYPE" = "zip" ]; then
        TARGET_ARCHIVE=dist/images/${APP_NAME}_$APP_VERSION.$TARGET_NAME.zip
        echo "Building compressed image $TARGET_ARCHIVE..."
        zip -vrj -FS $TARGET_ARCHIVE $TARGET_DIR -x "*.gitkeep" -x "*.DS_Store" > /dev/null
    else
        TARGET_ARCHIVE=dist/images/${APP_NAME}_$APP_VERSION.$TARGET_NAME.tgz
        echo "Building compressed image $TARGET_ARCHIVE..."
        $TAR_APP -czf $TARGET_ARCHIVE -C $BUILD_DIR_BASE --exclude=.gitkeep --exclude=.DS_Store $BUILD_DIR_FINAL
    fi
    echo ""
}

# Run builds for all target platforms
do_build darwin amd64 zip
do_build darwin arm64 zip
do_build linux amd64 tar
do_build linux arm tar

echo "${GREEN}Done.${ENDCOLOR}"


majortom.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"os/user"
	"path/filepath"
	"sort"
	"strings"

	"github.com/gookit/color"
)

const APP_NAME = "majortom"
const APP_VERSION = "1.4.0"

type ConfigDataT struct {
	Locations map[string]string `json:"locations"`
}
type ColorT struct {
	colorRGB color.RGBStyle
	color256 color.Color256
	color16  color.Color
}

var colorShortcut = ColorT{
	colorRGB: *color.HEXStyle("#ff8000"),
	color256: color.C256(208), // Orange
	color16:  color.Magenta,
}
var colorPath = ColorT{
	colorRGB: *color.HEXStyle("#00ffff"),
	color256: color.C256(87), // Cyan
	color16:  color.Cyan,
}
var colorPathDNE = ColorT{
	colorRGB: *color.HEXStyle("#808080"),
	color256: color.C256(246), // Gray
	color16:  color.Red,
}
var colorCurrent = ColorT{
	colorRGB: *color.HEXStyle("#ffff00"),
	color256: color.C256(190), // Yellow
	color16:  color.Yellow,
}
var colorError = ColorT{
	colorRGB: *color.HEXStyle("#ff4040"),
	color256: color.C256(198), // Red
	color16:  color.Red,
}

const (
	ColorMode16 = iota
	ColorMode256
	ColorMode16m
	ColorModeNone
)

var colorMode int = ColorMode16

var EXIT_CODE_SUCCESS = 0
var EXIT_CODE_FAIL = 1

var ENV_VAR_CONFIG = "MAJORTOM_CONFIG"
var DEFAULT_CONFIG_PATH = "~/.config/majortom/majortom_config.json"

func main() {
	flag.Usage = func() {
		w := flag.CommandLine.Output() // may be os.Stderr - but not necessarily
		executable := os.Args[0]
		fmt.Fprintf(w, "Usage of %s:\n", executable)
		fmt.Fprintf(w, "%s [-color=<color mode>|-no-color] <shortcut>\n", APP_NAME)
		fmt.Fprintf(w, "%s [-color=<color mode>|-no-color] [-a|-d] <shortcut>\n", APP_NAME)
		fmt.Fprintf(w, "%s -h\n", APP_NAME)
		fmt.Fprintf(w, "%s -init\n", APP_NAME)
		flag.PrintDefaults()
		fmt.Fprintf(
			w,
			"NOTE:\n"+
				"   majortom is meant to be invoked by the to() helper shell script function.\n"+
				"   Calling majortom directly will not change your working directory.\n"+
				"NOTE:\n"+
				"   Set the environment variable %s to point to %s's configuration file.\n"+
				"   If not set, the configuration file path will default to: %s\n"+
				"\n",
			ENV_VAR_CONFIG, APP_NAME, DEFAULT_CONFIG_PATH)

		configPath := os.Getenv(ENV_VAR_CONFIG)
		if configPath == "" {
			fmt.Fprintf(w, "Currently %s is not set.\n", ENV_VAR_CONFIG)
		} else {
			fmt.Fprintf(w, "Currently %s is set to \"%s\".\n", ENV_VAR_CONFIG, os.Getenv(ENV_VAR_CONFIG))
		}

		configPath = getConfigPath()
		fmt.Fprintf(w, "Expecting config file to be at: %s\n", configPath)
	}

	optVersion := flag.Bool("version", false,
		"Show version.")
	optAdd := flag.Bool("a", false,
		"Add current directory path as <shortcut>.")
	optDelete := flag.Bool("d", false,
		"Delete <shortcut>.")
	optInit := flag.Bool("init", false,
		"Initialize (create) config file. (Only if config does not exist)")
	optNoColor := flag.Bool("no-color", false,
		"Disable colorization")
	optColor := flag.String("color", "16m",
		"Set color mode. Can be set to any of: 16, 256, 16m.")
	flag.Parse()

	setColorMode(*optNoColor, *optColor)

	if *optVersion {
		// Show version and exit
		fmt.Printf("%s %s\n", APP_NAME, APP_VERSION)
		os.Exit(EXIT_CODE_SUCCESS)
	}

	if *optInit {
		// Initialize (create) config file and exit
		initConfig()
		os.Exit(EXIT_CODE_SUCCESS)
	}

	args := flag.Args()
	if len(args) > 1 {
		// Too many args.
		fmt.Println("Too many arguments.")
		os.Exit(EXIT_CODE_FAIL)
	}

	config := loadConfig()

	if len(args) == 0 {
		// With no arguments, show the current list of shortcuts and exit
		showShortcuts(config)
		os.Exit(EXIT_CODE_SUCCESS)
	}
	if *optAdd {
		// Add a new shortcut
		config = addShortcut(config, args[0])
		saveConfig(config)
		os.Exit(EXIT_CODE_SUCCESS)
	}
	if *optDelete {
		// Delete a shortcut
		config = deleteShortcut(config, args[0])
		saveConfig(config)
		os.Exit(EXIT_CODE_SUCCESS)
	}

	// Lookup the requested shortcut, and return it prepended by ":"
	path := getPath(config, args[0])
	path = expandHome(path)
	fmt.Printf(":%s\n", path)

	os.Exit(EXIT_CODE_SUCCESS)
}

func setColorMode(optNoColor bool, optColor string) {
	if optNoColor {
		colorMode = ColorModeNone
		return
	}
	switch optColor {
	case "16":
		colorMode = ColorMode16
	case "256":
		colorMode = ColorMode256
	case "16m":
		colorMode = ColorMode16m
	}
}

func addShortcut(config ConfigDataT, shortcut string) ConfigDataT {
	currentPath, err := os.Getwd()
	if err != nil {
		fmt.Println(err)
		os.Exit(EXIT_CODE_FAIL)
	}
	currentPath = abbreviateHome(currentPath)
	config.Locations[shortcut] = currentPath
	fmt.Printf("Adding shortcut \"%s\"...\n", shortcut)
	return config
}

func deleteShortcut(config ConfigDataT, shortcut string) ConfigDataT {
	if _, ok := config.Locations[shortcut]; ok {
		delete(config.Locations, shortcut)
	} else {
		colorPrintFLn(
			colorError,
			"Shortcut \"%s\" does not exist.",
			shortcut)
		os.Exit(EXIT_CODE_FAIL)
	}
	fmt.Printf("Deleting shortcut \"%s\"...\n", shortcut)
	return config
}

// Get the network path associated with the requested shortcut.
//
// - The shortcut can be abbreviated.
// - If no match is found then exit as failure.
// - If the requested shortcut does not unambiguously match a SINGLE defined shortcut, then
//   print a list of the matching shortcuts and exit as failure.
//
func getPath(config ConfigDataT, shortcut string) string {
	paths := make([]string, 0)
	matched_keys := make([]string, 0)

	for key, path := range config.Locations {
		if key == shortcut {
			// Always return the path if the key EXACTLY matches the requested shortcut
			return path
		}
		if strings.HasPrefix(key, shortcut) {
			paths = append(paths, path)
			matched_keys = append(matched_keys, key)
		}
	}
	if len(paths) == 0 {
		colorPrintFLn(
			colorError,
			"No match found for shortcut \"%s\". Run \"to\" with no arguments for a list of shortcuts.",
			shortcut)
		os.Exit(EXIT_CODE_FAIL)
	}
	if len(paths) > 1 {
		message := "Matched multiple shortcuts: "
		for i, key := range matched_keys {
			message += colorSprintF(colorShortcut, "%s", key)
			if i < len(matched_keys)-1 {
				message += ", "
			}
		}
		fmt.Println(message)
		os.Exit(EXIT_CODE_FAIL)
	}
	// Return the (single) matching path
	return paths[0]
}

// Print all configured shortcuts
func showShortcuts(config ConfigDataT) {
	currentPath, err := os.Getwd()
	if err != nil {
		fmt.Println(err)
		os.Exit(EXIT_CODE_FAIL)
	}

	var maxLen int
	shortcuts := make([]string, 0)
	for shortcut := range config.Locations {
		if len(shortcut) > maxLen {
			maxLen = len(shortcut)
		}
		shortcuts = append(shortcuts, shortcut)
	}

	if len(shortcuts) == 0 {
		fmt.Println(
			"No shortcuts exist yet.  Use the \"-a <shortcut>\" command " +
				"to create your first shortcut.")
		return
	}

	fmt.Println("Available shortcuts:")
	sort.Strings(shortcuts)
	for _, shortcut := range shortcuts {
		path := config.Locations[shortcut]
		pathAbsolute := expandHome(path)
		if pathAbsolute == currentPath {
			colorPrintF(colorCurrent, "▶ %-*s ", maxLen, shortcut)
		} else {
			colorPrintF(colorShortcut, "  %-*s ", maxLen, shortcut)
		}
		if _, err := os.Stat(pathAbsolute); !os.IsNotExist(err) {
			// Path exists
			colorPrintFLn(colorPath, "%s", path)
		} else {
			// Path does not exist
			if colorMode == ColorModeNone {
				fmt.Printf("%s (DOES NOT EXIST)\n", path)
			} else {
				colorPrintFLn(colorPathDNE, "%s", path)
			}
		}
	}
}

// Replace '~' in requested path with the current user's home path.
func expandHome(path string) string {
	usr, _ := user.Current()
	homeDir := usr.HomeDir

	if path == "~" {
		// In case of "~", which won't be caught by the "else if"
		return homeDir
	} else if strings.HasPrefix(path, "~/") {
		// Use strings.HasPrefix so we don't match paths like
		// "/something/~/something/"
		return filepath.Join(homeDir, path[2:])
	}
	return path
}

// Replace the current user's home path within the requested path with `~`.
func abbreviateHome(path string) string {
	usr, _ := user.Current()
	dir := usr.HomeDir
	return strings.Replace(path, dir, "~", 1)
}

// Load the config file and return it as a ConfigDataT object.
// Exits on failure.
func loadConfig() ConfigDataT {
	pathConfig := getConfigPath()

	if _, err := os.Stat(pathConfig); os.IsNotExist(err) {
		colorPrintFLn(
			colorError,
			"The %s config file (%s) does not exist.  You can create it using the -init command.",
			APP_NAME,
			pathConfig)
		os.Exit(EXIT_CODE_FAIL)
	}

	jsonFile, err := os.Open(pathConfig)
	if err != nil {
		fmt.Println(err)
		os.Exit(EXIT_CODE_FAIL)
	}
	defer jsonFile.Close()

	var config ConfigDataT
	byteValue, _ := ioutil.ReadAll(jsonFile)
	json.Unmarshal(byteValue, &config)
	return config
}

func saveConfig(config ConfigDataT) {
	pathConfig := getConfigPath()
	file, _ := json.MarshalIndent(config, "", "    ")
	_ = ioutil.WriteFile(pathConfig, file, 0644)
}

// Initialize the configuration file (IF it does not already exist).
//
// This function will NEVER overwrite an existing config file.
//
func initConfig() {
	pathConfig := getConfigPath()

	if _, err := os.Stat(pathConfig); !os.IsNotExist(err) {
		colorPrintFLn(
			colorError,
			"The %s config file (%s) already exists. The -init option cannot be used to reinitialize"+
				" an existing config file. If you really want to re-create the config file then"+
				" delete it manually and re-run this command.",
			APP_NAME,
			pathConfig)
		os.Exit(EXIT_CODE_FAIL)
	}

	// Make a blank config struct
	config := ConfigDataT{}
	config.Locations = map[string]string{}

	// Create missing config directory (and its parents), if they don't exist
	pathConfig = getConfigPath()
	parent := filepath.Dir(pathConfig)
	if _, err := os.Stat(parent); os.IsNotExist(err) {
		// parent does not exist
		fmt.Printf("Creating directory: %s\n", parent)
		err := os.MkdirAll(parent, 0755)
		if err != nil {
			fmt.Println(err)
			os.Exit(EXIT_CODE_FAIL)
		}
	}

	saveConfig(config)
	fmt.Printf(
		"A new %s config file has been initialized (created) at:\n   %s\n",
		APP_NAME,
		pathConfig)
}

// Get the path of the current config file (with ~ expanded).
func getConfigPath() string {
	path := os.Getenv(ENV_VAR_CONFIG)
	if path == "" {
		path = DEFAULT_CONFIG_PATH
	}
	path = expandHome(path)
	return path
}

// Print a colorized formatted string
func colorPrintF(textColor ColorT, format string, args ...interface{}) {
	switch colorMode {
	case ColorMode16m:
		textColor.colorRGB.Printf(format, args...)
	case ColorMode16:
		textColor.color16.Printf(format, args...)
	case ColorMode256:
		textColor.color256.Printf(format, args...)
	case ColorModeNone:
		fmt.Printf(format, args...)
	}
}

// Print a colorized formatted string, with a terminating linefeed AFTER the color reset escape code.
//
// The output of this application is captured and echoed by a shell script, and the shell doesn't
// recognize that echo'd content ends in a linefeed IF that linefeed occurs BEFORE the
// escape codes used to clear the color.  To fix that issue we use this wrapper function which
// will inject a terminating linefeed AFTER the color reset escape code.
//
func colorPrintFLn(textColor ColorT, format string, args ...interface{}) {
	switch colorMode {
	case ColorMode16m:
		fmt.Printf("%s\n", textColor.colorRGB.Sprintf(format, args...))
	case ColorMode16:
		fmt.Printf("%s\n", textColor.color16.Sprintf(format, args...))
	case ColorMode256:
		fmt.Printf("%s\n", textColor.color256.Sprintf(format, args...))
	case ColorModeNone:
		fmt.Printf("%s\n", fmt.Sprintf(format, args...))
	}

}

// String print a colorized formatted string
func colorSprintF(textColor ColorT, format string, args ...interface{}) string {
	switch colorMode {
	case ColorMode16m:
		return textColor.colorRGB.Sprintf(format, args...)
	case ColorMode16:
		return textColor.color16.Sprintf(format, args...)
	case ColorMode256:
		return textColor.color256.Sprintf(format, args...)
	default:
		return fmt.Sprintf(format, args...)
	}
}
